src/driver-steelseries.c

```diff
@@ -440,98 +512,72 @@ steelseries_write_led_v2(struct ratbag_led *led)
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point[0].color = black;
+		point[0].pos = 0x00;
+
+		list_append(&cycle.points, &point[0].link);
 		break;
 	case RATBAG_LED_ON:
-		buf[19] = 0x01;
-		buf[27] = 0x01;
-		buf[28] = buf[31] = led->color.red;
-		buf[29] = buf[32] = led->color.green;
-		buf[30] = buf[33] = led->color.blue;
-
-		/* not sure why the duration is set for the steady color or why
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point[0].color = led->color;
+
+		list_append(&cycle.points, &point[0].link);
 		break;
-	case RATBAG_LED_CYCLE:
-		buf[27] = 0x04; /* number of steps in cycle */
-
-		/* start color */
-		buf[28] = buf[31] = 0xFF;
-		buf[29] = buf[32] = 0x00;
-		buf[30] = buf[33] = 0x00;
-
-		/* Cycle to green */
-		buf[35] = 0x00;
-		buf[36] = 0xFF;
-		buf[37] = 0x00;
-		buf[38] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to blue */
-		buf[39] = 0x00;
-		buf[40] = 0x00;
-		buf[41] = 0xFF;
-		buf[42] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to red */
-		buf[43] = 0xFF;
-		buf[44] = 0x00;
-		buf[45] = 0x00;
-		buf[46] = 0x56; /* normalized time share of animation */
-
-		duration = led->ms;
+	case RATBAG_LED_CYCLE:		
+		point[0].color = red;
+		point[0].pos = 0x00;
+
+		point[1].color = green;
+		point[1].pos = 0x55;
+
+		point[2].color = blue;
+		point[2].pos = 0x55;
+
+		point[3].color = red;
+		point[3].pos = 0x55;
+
+		list_append(&cycle.points, &point[0].link);
+		list_append(&cycle.points, &point[1].link);
+		list_append(&cycle.points, &point[2].link);
+		list_append(&cycle.points, &point[3].link);
+
+		cycle.duration = led->ms;
 		break;
 	case RATBAG_LED_BREATHING:
-		buf[27] = 0x03; /* number of steps in cycle */
-
-		/* start color */
-		buf[28] = buf[31] = led->color.red;
-		buf[29] = buf[32] = led->color.green;
-		buf[30] = buf[33] = led->color.blue;
-
-		/* Cycle to black */
-		buf[35] = 0x00;
-		buf[36] = 0x00;
-		buf[37] = 0x00;
-		buf[38] = 0x7F; /* normalized time share of animation */
-
-		/* Cycle to selected color */
-		buf[39] = led->color.red;
-		buf[40] = led->color.green;
-		buf[41] = led->color.blue;
-		buf[42] = 0x7F; /* normalized time share of animation */
-
-		duration = led->ms;
+		point[0].color = black;
+		point[0].pos = 0x00;
+
+		point[1].color = led->color;
+		point[1].pos = 0x7F;
+
+		point[2].color = black;
+		point[2].pos = 0x7F;
+		
+		list_append(&cycle.points, &point[0].link);
+		list_append(&cycle.points, &point[1].link);
+		list_append(&cycle.points, &point[2].link);
+
+		cycle.duration = led->ms;
 		break;
 	default:
 		return -EINVAL;
 	}
 
-	/* this seems to be the minimum allowed */
-	min_duration = buf[27] * 330;
-	if (duration < min_duration)
-		duration = min_duration;
-
-	hidpp_set_unaligned_le_u16(&buf[3], duration);
+	construct_cycle_buffer(&cycle, buf, STEELSERIES_REPORT_SIZE);
```

Comment:
sizeof(buf) would make this more flexible if we have to use a different report size in the future